apps/api/src/app/workflows-v2/usecases/generate-preview/generate-preview.usecase.ts

```diff
@@ -241,3 +334,171 @@ class FrameworkError {
   message: string;
   name: string;
 }
+
+/**
+ * Fixes invalid control values by applying default values from the schema
+ *
+ * @example
+ * // Input:
+ * const values = { foo: 'invalid' };
+ * const errors = [{ instancePath: '/foo' }];
+ * const defaults = { foo: 'default' };
+ *
+ * // Output:
+ * const fixed = { foo: 'default' };
+ */
+function replaceInvalidControlValues(
+  normalizedControlValues: Record<string, unknown>,
+  errors: ErrorObject[]
+): Record<string, unknown> {
+  const fixedValues = _.cloneDeep(normalizedControlValues);
+
+  errors.forEach((error) => {
+    const path = getErrorPath(error);
+    const defaultValue = _.get(previewControlValueDefault, path);
+    _.set(fixedValues, path, defaultValue);
+  });
+
+  return fixedValues;
+}
+
+function processControlValuesBySchema(
+  controlSchema: JSONSchemaDto | undefined,
+  controlValues: Record<string, unknown>,
+  stepData: StepDataDto
+): { issues: StepIssuesDto; schemaCompliantControls: Record<string, unknown> } {
+  const issues: StepIssuesDto = {};
+  let fixedValues: Record<string, unknown> | undefined;
+
+  const schemaCompliantControls = normalizeControlValues(controlValues, stepData.type);
+
+  if (!schemaCompliantControls) {
+    throw new Error(
+      // eslint-disable-next-line max-len
+      'Control values normalization failed: The normalizeControlValues function requires maintenance to handle the provided type or data structure correctly'
+    );
+  }
+
+  /*
+   * TODO: Consider if this validation is necessary since we already normalize in normalizeControlValues and,
+   * we can make it more robust by adding custom parsable by liquid js values
+   *
+   * Current purposes:
+   * 1. Validates control schema for code-first approach
+   * 2. Allows adding custom validation rules
+   *    Example: If emailEditor is renamed to 'body', we can validate it's valid TipTap JSON
+   */
+  if (controlSchema) {
+    const ajv = new Ajv({ allErrors: true });
+    addFormats(ajv);
+
+    const validate = ajv.compile(controlSchema);
+    const isValid = validate(schemaCompliantControls);
+    const errors = validate.errors as null | ErrorObject[];
+
+    if (!isValid && errors && errors?.length !== 0 && schemaCompliantControls) {
+      fixedValues = replaceInvalidControlValues(schemaCompliantControls, errors);
+
+      return { issues, schemaCompliantControls: fixedValues };
+    }
+  }
```

Comment:
this part, with the help of strict JSON schema/Zod, will work well for generating issues. However, in this context, I believe we can focus on improving `normalizeControlValues` to handle the heavy lifting. At this stage, the control values would then be clean and valid (aligned with the control schema and framework output schema), ensuring the preview runs smoothly.

Reply:
please disregard the logic in the `if (controlSchema)`  statement (with the zod validation and replace by issues), as I removed it in favor of utilizing `normalizeControlValues` more effectively, as explained earlier.